tasks/task_translation.c

```diff
@@ -891,275 +1265,488 @@ bool run_translation_service(settings_t *settings, bool paused)
+         
+         for (i = 0; i < ARRAY_SIZE(ACCESS_INPUT_LABELS); i++)
          {
-            case 2:
-               strlcpy(new_ai_service_url       + _len,
-                     "text",
-                     sizeof(new_ai_service_url) - _len);
-               break;
-            case 1:
-            case 3:
-               _len += strlcpy(new_ai_service_url    + _len,
-                     "sound,wav",
-                     sizeof(new_ai_service_url)      - _len);
-               if (ai_service_mode == 1)
-                  break;
-               /* fall-through intentional for ai_service_mode == 3 */
-            case 0:
-               _len += strlcpy(new_ai_service_url    + _len,
-                     "image,png",
-                     sizeof(new_ai_service_url)      - _len);
+            rjsonwriter_add_comma(writer);
+            rjsonwriter_add_string(writer, ACCESS_INPUT_LABELS[i]);
+            rjsonwriter_add_colon(writer);
+            rjsonwriter_add_unsigned(writer, request->inputs[i]);
+         }      
+         rjsonwriter_add_end_object(writer);
+      }
+      rjsonwriter_add_end_object(writer);
+   }
+ 
+   return writer;
+}
+
+/**
+ * Writes in the provided {buffer} the URL for the translation request. The 
+ * buffer is guaranteed to contain the server URL as well as an 'output' param
+ * specifying the accepted data types for this service.
+ */
+static void build_request_url(char *buffer, size_t length, settings_t *settings)
+{
+   char token[2];
+   bool poke_supported              = false;
+   unsigned service_source_lang     = settings->uints.ai_service_source_lang;
+   unsigned service_target_lang     = settings->uints.ai_service_target_lang;
+   const char *service_url          = settings->arrays.ai_service_url;
+   unsigned ai_service_mode         = settings->uints.ai_service_mode;
 #ifdef HAVE_GFX_WIDGETS
-               if (     video_st->poke
-                     && video_st->poke->load_texture
-                     && video_st->poke->unload_texture)
-                  strlcpy(new_ai_service_url       + _len,
-                        ",png-a",
-                        sizeof(new_ai_service_url) - _len);
+   video_driver_state_t *video_st   = video_state_get_ptr();
+   poke_supported                   = video_st->poke
+                                   && video_st->poke->load_texture
+                                   && video_st->poke->unload_texture;
 #endif
-               break;
-            default:
-               break;
-         }
+      
+   strlcpy(buffer, service_url, length);
+   
+   token[1] = '\0';
+   if (strrchr(buffer, '?'))
+      token[0] = '&';
+   else
+      token[0] = '?';
+   
+   if (service_source_lang != TRANSLATION_LANG_DONT_CARE)
+   {
+      const char *lang_source 
+            = ai_service_get_str((enum translation_lang)service_source_lang);
 
+      if (!string_is_empty(lang_source))
+      {
+         strlcat(buffer, token, length);
```

Comment:
Can you find a way to do away with all these strlcat calls and instead use strlcpy? We want to avoid use of strlcpy as much as possible. It is possible to concatenate a string with strlcpy too (see prior examples in other RetroArch files)

Reply:
[
![1](https://github.com/libretro/RetroArch/assets/12669784/cdfbfa7f-31a6-45ff-8e0e-6b0cc3180c4a)
Would this be the correct way to replace it? Would I need to keep checking the buffer length?

Reply:
Please avoid strlens entirely too. I want strlens avoided like the absolute plague and only used when ABSOLUTELY necessary and no other alternative is possible.

No strlens please and no strlcats either. Just look at prior examples of how this is done. Do a search through the codebase, grep for 'strlcpy'.

Reply:
Here is a good example of how to do all this without strlens and without strlcats -

https://github.com/libretro/RetroArch/blob/master/libretro-common/file/file_path.c#L338

There are other examples like this sprinkled all throughout the codebase.

Reply:
Ok thank you.  Would this be better?
``
_len  = strlcpy(buffer, token, length);
 _len += strlcpy(buffer + _len, "source_lang=", length - _len);
strlcpy(buffer + _len, lang_source, length - _len);
``

Reply:
@Cpod12 Ye, that example would be sufficient.

Reply:
Ok it should be good now